dbms/src/Dictionaries/PolygonDictionary.cpp

```diff
@@ -0,0 +1,747 @@
+            this->database,
+            this->name,
+            this->dict_struct,
+            this->source_ptr->clone(),
+            this->dict_lifetime,
+            this->input_type,
+            this->point_type);
+}
+
+bool SimplePolygonDictionary::find(const Point &point, size_t & id) const
+{
+    bool found = false;
+    double area = 0;
+    for (size_t i = 0; i < (this->polygons).size(); ++i)
+    {
+        if (bg::covered_by(point, (this->polygons)[i]))
+        {
+            double new_area = bg::area((this->polygons)[i]);
+            if (!found || new_area < area)
+            {
+                found = true;
+                id = i;
+                area = new_area;
+            }
+        }
+    }
+    return found;
+}
+
+void registerDictionaryPolygon(DictionaryFactory & factory)
+{
+    auto create_layout = [=](const std::string &,
+                             const DictionaryStructure & dict_struct,
+                             const Poco::Util::AbstractConfiguration & config,
+                             const std::string & config_prefix,
+                             DictionarySourcePtr source_ptr) -> DictionaryPtr
+    {
+        const String database = config.getString(config_prefix + ".database", "");
+        const String name = config.getString(config_prefix + ".name");
+
+        if (!dict_struct.key)
+            throw Exception{"'key' is required for a dictionary of layout 'polygon'", ErrorCodes::BAD_ARGUMENTS};
+        if (dict_struct.key->size() != 1)
+            throw Exception{"The 'key' should consist of a single attribute for a dictionary of layout 'polygon'",
+                            ErrorCodes::BAD_ARGUMENTS};
+        // TODO: Once arrays are fully supported this should be changed to a more reasonable check.
+        const auto type_name = (*dict_struct.key)[0].type->getName();
+        IPolygonDictionary::InputType input_type;
+        IPolygonDictionary::PointType point_type;
+        if (type_name == "Array(Array(Array(Array(Float64))))")
+        {
+            input_type = IPolygonDictionary::InputType::MultiPolygon;
+            point_type = IPolygonDictionary::PointType::Array;
+        }
+        else if (type_name == "Array(Array(Array(Tuple(Float64, Float64))))")
+        {
+            input_type = IPolygonDictionary::InputType::MultiPolygon;
+            point_type = IPolygonDictionary::PointType::Tuple;
+        }
+        else if (type_name == "Array(Array(Float64))")
+        {
+            input_type = IPolygonDictionary::InputType::SimplePolygon;
+            point_type = IPolygonDictionary::PointType::Array;
+        }
+        else if (type_name == "Array(Tuple(Float64, Float64))")
+        {
+            input_type = IPolygonDictionary::InputType::SimplePolygon;
+            point_type = IPolygonDictionary::PointType::Tuple;
+        }
+        else
+            throw Exception{"The key type is not one of the allowed types for a dictionary of layout 'polygon'",
```

Comment:
To make exception message more user-friendly, you can add to it:
* current key type
* key types which are supported

Reply:
Done.